Import Libraries

In [1]:
import pandas as pd
import numpy as np


#### Q1. Pandas version
What's the version of Pandas that you installed?

In [2]:
pd.__version__

'2.3.1'

Downloading the the Car Fuel Efficiency dataset

In [3]:
df = pd.read_csv('https://raw.githubusercontent.com/alexeygrigorev/datasets/master/car_fuel_efficiency.csv')

In [4]:
#viewing the first few rows of dataset
df.head()

engine_displacement  num_cylinders  horsepower  vehicle_weight  \
0                  170            3.0       159.0     3413.433759   
1                  130            5.0        97.0     3149.664934   
2                  170            NaN        78.0     3079.038997   
3                  220            4.0         NaN     2542.392402   
4                  210            1.0       140.0     3460.870990   

   acceleration  model_year  origin fuel_type         drivetrain  num_doors  \
0          17.7        2003  Europe  Gasoline    All-wheel drive        0.0   
1          17.8        2007     USA  Gasoline  Front-wheel drive        0.0   
2          15.1        2018  Europe  Gasoline  Front-wheel drive        0.0   
3          20.2        2009     USA    Diesel    All-wheel drive        2.0   
4          14.4        2009  Europe  Gasoline    All-wheel drive        2.0   

   fuel_efficiency_mpg  
0            13.231729  
1            13.688217  
2            14.246341  
3            16.912736  
4            12.488369

#### Q2. Records count
How many records are in the dataset?

In [6]:
df.shape

(9704, 11)

In this dataset there is 9704 rows and 11 columns

In [8]:
# Structure of dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9704 entries, 0 to 9703
Data columns (total 11 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   engine_displacement  9704 non-null   int64  
 1   num_cylinders        9222 non-null   float64
 2   horsepower           8996 non-null   float64
 3   vehicle_weight       9704 non-null   float64
 4   acceleration         8774 non-null   float64
 5   model_year           9704 non-null   int64  
 6   origin               9704 non-null   object 
 7   fuel_type            9704 non-null   object 
 8   drivetrain           9704 non-null   object 
 9   num_doors            9202 non-null   float64
 10  fuel_efficiency_mpg  9704 non-null   float64
dtypes: float64(6), int64(2), object(3)
memory usage: 834.1+ KB


#### Q3. Fuel types
How many fuel types are presented in the dataset?

In [9]:
 df.fuel_type.nunique()

2

### Q4. Missing values
How many columns in the dataset have missing values?

In [10]:
df.isnull().sum()

engine_displacement      0
num_cylinders          482
horsepower             708
vehicle_weight           0
acceleration           930
model_year               0
origin                   0
fuel_type                0
drivetrain               0
num_doors              502
fuel_efficiency_mpg      0
dtype: int64

In [11]:
num_col_missing_values = (df.isnull().sum() > 0).sum().item()
num_col_missing_values

4

In [12]:
#statistical summary of numeric columns
df.describe()

engine_displacement  num_cylinders   horsepower  vehicle_weight  \
count          9704.000000    9222.000000  8996.000000     9704.000000   
mean            199.708368       3.962481   149.657292     3001.280993   
std              49.455319       1.999323    29.879555      497.894860   
min              10.000000       0.000000    37.000000      952.681761   
25%             170.000000       3.000000   130.000000     2666.248985   
50%             200.000000       4.000000   149.000000     2993.226296   
75%             230.000000       5.000000   170.000000     3334.957039   
max             380.000000      13.000000   271.000000     4739.077089   

       acceleration   model_year    num_doors  fuel_efficiency_mpg  
count   8774.000000  9704.000000  9202.000000          9704.000000  
mean      15.021928  2011.484027    -0.006412            14.985243  
std        2.510339     6.659808     1.048162             2.556468  
min        6.000000  2000.000000    -4.000000             6.200971  
25%       13.300000  2006.000000    -1.000000            13.267459  
50%       15.000000  2012.000000     0.000000            15.006037  
75%       16.700000  2017.000000     1.000000            16.707965  
max       24.300000  2023.000000     4.000000            25.967222

#### Q5. Max fuel efficiency
What's the maximum fuel efficiency of cars from Asia?

In [15]:
# Filter rows where origin is Asia and get max fuel efficiency
max_asia = df.loc[df["origin"] == "Asia", "fuel_efficiency_mpg"].max()
print("Max fuel efficiency of cars from Asia:", round(max_asia,3))

Max fuel efficiency of cars from Asia: 23.759


#### Q6. Median value of horsepower

In [18]:
#Find the median value of horsepower column in the dataset.
print("Median horsepower:", df["horsepower"].median())

Median horsepower: 149.0


In [21]:
#calculate the most frequent value of the same horsepower column.
most_frequent_hp = df["horsepower"].mode()[0]
print("Most frequent horsepower:", most_frequent_hp)

Most frequent horsepower: 152.0


In [22]:
# Use fillna method to fill the missing values in horsepower column with the mode
df["horsepower"] = df["horsepower"].fillna(most_frequent_hp)

In [23]:
# verifying are there any missing values
print(df["horsepower"].isnull().sum())


0


In [24]:
# calculate the median value of horsepower once again.
print("New Median horsepower:", df["horsepower"].median())

New Median horsepower: 152.0


#### Q7. Sum of weights

In [29]:
# Select all the cars from Asia
cars_asia = df[df["origin"] == "Asia"]
print(cars_asia.head())
print("Number of Asian cars:", len(cars_asia))

    engine_displacement  num_cylinders  horsepower  vehicle_weight  \
8                   250            1.0       174.0     2714.219310   
12                  320            5.0       145.0     2783.868974   
14                  200            6.0       160.0     3582.687368   
20                  150            3.0       197.0     2231.808142   
21                  160            4.0       133.0     2659.431451   

    acceleration  model_year origin fuel_type         drivetrain  num_doors  \
8           10.3        2016   Asia    Diesel  Front-wheel drive       -1.0   
12          15.1        2010   Asia    Diesel    All-wheel drive        1.0   
14          14.9        2007   Asia    Diesel    All-wheel drive        0.0   
20          18.7        2011   Asia  Gasoline  Front-wheel drive        1.0   
21           NaN        2016   Asia  Gasoline  Front-wheel drive       -1.0   

    fuel_efficiency_mpg  
8             16.823554  
12            16.175820  
14            11.871091  


In [27]:
# Select only columns vehicle_weight and model_year
df_cars_asia = cars_asia[["vehicle_weight" ,"model_year"]]
df_cars_asia.head()

vehicle_weight  model_year
8      2714.219310        2016
12     2783.868974        2010
14     3582.687368        2007
20     2231.808142        2011
21     2659.431451        2016

In [30]:
#Select the first 7 values
first_seven = df_cars_asia.head(7)
print(first_seven)

    vehicle_weight  model_year
8      2714.219310        2016
12     2783.868974        2010
14     3582.687368        2007
20     2231.808142        2011
21     2659.431451        2016
34     2844.227534        2014
38     3761.994038        2019


In [31]:
# Get the underlying NumPy array. Let's call it X
X = first_seven.to_numpy()
print(X)

[[2714.21930965 2016.        ]
 [2783.86897424 2010.        ]
 [3582.68736772 2007.        ]
 [2231.8081416  2011.        ]
 [2659.43145076 2016.        ]
 [2844.22753389 2014.        ]
 [3761.99403819 2019.        ]]


In [33]:
XTX = X.T @ X 
XTX_inv = np.linalg.inv(XTX)


In [35]:
#Create an array
y = np.array([1100, 1300, 800, 900, 1000, 1100, 1200])
print(y)

[1100 1300  800  900 1000 1100 1200]


In [36]:
w = XTX_inv @ X.T @ y
print("Weights (w):", w)

Weights (w): [0.01386421 0.5049067 ]


In [37]:
sum_w = np.sum(w)
print("Sum of all elements in w:", sum_w)

Sum of all elements in w: 0.5187709081074016
